src/audio/codec_adapter/codec_adapter.c

```diff
@@ -0,0 +1,528 @@
+			goto end;
+		} else if (!size) {
+			comp_err(dev, "codec_adapter_set_params() error: no configuration size %d",
+				 size);
+			//TODO: return -EINVAL. This is temporary until driver fixes its issue
+			ret = 0;
+			goto end;
+		} else if (size > MAX_BLOB_SIZE) {
+			comp_err(dev, "codec_adapter_set_params() error: blob size is too big cfg size %d, allowed %d",
+				 size, MAX_BLOB_SIZE);
+			ret = -EINVAL;
+			goto end;
+		} else if (type != CODEC_CFG_SETUP && type != CODEC_CFG_RUNTIME) {
+			comp_err(dev, "codec_adapter_set_params() error: unknown config type");
+			ret = -EINVAL;
+			goto end;
+		}
+
+		/* Allocate buffer for new params */
+		cd->runtime_params = rballoc(0, SOF_MEM_CAPS_RAM, size);
+		if (!cd->runtime_params) {
+			comp_err(dev, "codec_adapter_set_params(): space allocation for new params failed");
+			ret = -ENOMEM;
+			goto end;
+		}
+
+		memset(cd->runtime_params, 0, size);
+	} else if (!cd->runtime_params) {
+		comp_err(dev, "codec_adapter_set_params() error: no memory available for runtime params in consecutive load");
+		ret = -EIO;
+		goto end;
+	}
+
+	offset = size - (cdata->num_elems + cdata->elems_remaining);
+	dst = (char *)cd->runtime_params + offset;
+	src = (char *)cdata->data->data;
+
+	ret = memcpy_s(dst, size - offset, src, cdata->num_elems);
+	assert(!ret);
+
+	/* Config has been copied now we can load & apply it depending on
+	 * codec state.
+	 */
+	if (!cdata->elems_remaining) {
+		switch (type) {
+		case CODEC_CFG_SETUP:
+			ret = load_setup_config(dev, cd->runtime_params, size);
+			if (ret) {
+				comp_err(dev, "codec_adapter_set_params(): error %d: load of setup config failed.",
+					 ret);
+			} else {
+				comp_dbg(dev, "codec_adapter_set_params() load of setup config done.");
+			}
+
+			break;
+		case CODEC_CFG_RUNTIME:
+			break;
+		default:
+			comp_err(dev, "codec_adapter_set_params(): error: unknown config type.");
+			break;
+		}
+	}
+end:
+	return ret;
+}
+
+static int ca_set_binary_data(struct comp_dev *dev,
+			      struct sof_ipc_ctrl_data *cdata)
+{
+	int ret;
+
+	comp_info(dev, "ca_set_binary_data() start, data type %d",
+		  cdata->data->type);
+
+	switch (cdata->data->type) {
+	case CODEC_CFG_SETUP:
+	case CODEC_CFG_RUNTIME:
+		ret = codec_adapter_set_params(dev, cdata, cdata->data->type);
+		break;
+	default:
+		comp_err(dev, "ca_set_binary_data() error: unknown binary data type");
+		ret = -EINVAL;
+		break;
+	}
+
+	return ret;
+}
+
+static int codec_adapter_ctrl_set_data(struct comp_dev *dev,
+				       struct sof_ipc_ctrl_data *cdata)
+{
+	int ret;
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	comp_info(dev, "codec_adapter_ctrl_set_data() start, state %d, cmd %d",
+		  cd->state, cdata->cmd);
+
+	/* Check version from ABI header */
+	if (SOF_ABI_VERSION_INCOMPATIBLE(SOF_ABI_VERSION, cdata->data->abi)) {
+		comp_err(dev, "codec_adapter_ctrl_set_data(): ABI mismatch!");
+		return -EINVAL;
+	}
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_ENUM:
+		//TODO
```

Comment:
?

Reply:
There is no need for `SOF_CTRL_CMD_ENUM` so I have added a `TODO` here. I can replace it with a error trace as well.

Reply:
error is good. Lets everyone know it's not implemented.